Add 'refresh', 'pause', 'resume', 'ps', and 'copy-image' commands to hack/dind-cluster.sh

knobunc opened the pull request:
The new refresh command will redeploy the binaries to a dind cluster
and restart all openshift processes that use the binaries so that you
can more easily test changes to openshift in a dind environment.

The pause command will stop the running containers in the cluster, but
not remove them.

The resume command will start containers in the cluster if they were
previously paused, or otherwise stopped (perhaps from a reboot).  It
tolerates when the containers are started with different IP addresses
by removing the appropriate state so the containers regenerate the
correct config when they start.

The new 'ps' command lists the status of all docker containers in the
cluster.
    
'copy-image' loads an image from the host docker into all node docker
daemons.
    
Also did some clean-up to make the function naming style consistent
and to add safety checks to all of the commands.

stevekuznetsov commented on hack/dind-cluster.sh:
```diff
@@ -160,6 +160,59 @@ function stop() {
+  echo "  Stopping master and node processes in cluster '${cluster_id}'..."
+  local master_cid
+  master_cid="$(${DOCKER_CMD} ps -qa --filter "name=${MASTER_NAME}")"
+  if [[ "${master_cid}" ]]; then
```
use `[[ -n "${var}" ]]` to explicitly test for it being set

stevekuznetsov commented on hack/dind-cluster.sh:
```diff
@@ -160,6 +160,59 @@ function stop() {
+  node_count=0
+  if [[ "${node_cids}" ]]; then
+    node_cids=(${node_cids//\n/ })
+    for cid in "${node_cids[@]}"; do
```
does 

```sh
for cid in $( ${DOCKER_CMD} ps -qa --filter "name=${NODE_PREFIX}" ); do

done
```

get you what you want? it will break on spaces and newlines by default

knobunc commented:
Thanks for the review @stevekuznetsov.  I'll investigate your suggestion.

knobunc commented:
@stevekuznetsov updated as requested... and other related cleanups.

stevekuznetsov commented:
@knobunc maybe a mistaken push? The diff doesn't look different on my end.

knobunc commented:
No... I am an idiot and forgot to add the change :-)  Sorry!

This version has two new commands 'pause' and 'resume' that stop and start the docker containers.  But stop and start were already taken... so I have the slightly confusing names.

stevekuznetsov commented on hack/dind-cluster.sh:
```diff
@@ -131,20 +131,13 @@ function stop() {
-  if [[ "${master_cid}" ]]; then
-    ${DOCKER_CMD} rm -f "${master_cid}" > /dev/null
-  fi
+  for cid in $( ${DOCKER_CMD} ps -qa --filter "name=${MASTER_NAME}" ); do
```
If you're doing the same thing to the master and node containers, the filter supports regex:

```sh
for cid in $( ${DOCKER_CMD} ps -qa --filter "name=${MASTER_NAME}|${NODE_PREFIX}" ); do
    #  ...
done
```

stevekuznetsov commented on hack/dind-cluster.sh:
```diff
@@ -160,6 +153,89 @@ function stop() {
 }
 
+function pause() {
+  local config_root=$1
```
Unused?

stevekuznetsov commented on hack/dind-cluster.sh:
```diff
@@ -160,6 +153,89 @@ function stop() {
+  local command=("$@")
+
+  for cid in $( ${DOCKER_CMD} ps -qa --filter "name=${filter}" ); do
+    ${DOCKER_CMD} exec "${cid}" ${command[@]}
```
Quote array expansions or they'll split on spaces not not on items

imcsk8 commented:
Code looks good. Testing manually

stevekuznetsov commented on hack/dind-cluster.sh:
```diff
@@ -212,17 +284,25 @@ function copy-runtime() {
+function wait-for-master() {
+  local config_root=$1
+
+  local kubeconfig="$(get-admin-config "${config_root}")"
```
If you mix scoping statements (`local`, `export`, `readonly`, `declare`, etc) and other constructs with return codes (like subshells) the return code of the scoping statement will alias that of the expression and if the expression fails this line will still succeed. So, we try to break them up:

```sh
local kubeconfig; kubeconfig="$( get-admin-config "${config_root}" )"
```

knobunc commented:
@stevekuznetsov @imcsk8 thanks for the reviews.  I pushed another update that addressed your concerns and also corrected the resume behavior so it tolerates IP addresses changing.

knobunc commented:
[test]

knobunc commented:
[testextended][extended:networking]

knobunc commented:
[test] [testextended][extended:networking]

knobunc commented:
[test]

danwinship commented on hack/dind-cluster.sh:
```diff
@@ -42,8 +42,9 @@
-# existing cluster is cleaned up before deploying a new cluster.
+# breaking golang changes, the -r argument to the start command will
+# ensure that an existing cluster is cleaned up before deploying a new
+# cluster.
```
can you split the fixes of existing code/comments into a separate commit?

danwinship commented:
> The new refresh command will redeploy the binaries to a dind cluster
> and restart all openshift processes that use the binaries so that you
> can more easily test changes to openshift in a dind environment.

Is this really that much faster than "start -r" to justify the additional code?

> The pause command will stop the running containers in the cluster, but
> not remove them.

What's the use case for this? If you're only going to be pausing it for a short period of time, then do you really need to bother pausing it, and if you're going to be pausing it for a longer period of time, then (a) will you actually want the old state when you come back or are you going to end up "refresh"ing anyway, and (b) will you actually remember what the old state is?

At least, maybe "resume" should warn you if the cluster currently has a build that doesn't match the currently-checked-out branch?

> The new 'ps' command lists the status of all docker containers in the
> cluster.

what's the use case for that?

> 'move-image' loads an image from the host docker into all node docker
> daemons.

Nitpick, but shouldn't that be "copy-image"?

knobunc commented:
> > The new refresh command will redeploy the binaries to a dind cluster
> > and restart all openshift processes that use the binaries so that you
> > can more easily test changes to openshift in a dind environment.

> Is this really that much faster than "start -r" to justify the additional code?

It's not the speed, it's the state that I care about.  e.g. Aaron and I are using this to deploy new OpenShift binaries to test rule changes without disturbing the two running pods.

> > The pause command will stop the running containers in the cluster, but
> > not remove them.

> What's the use case for this? If you're only going to be pausing it for a short period of time, then do you really need to bother pausing it, and if you're going to be pausing it for a longer period of time, then (a) will you actually want the old state when you come back or are you going to end up "refresh"ing anyway, and (b) will you actually remember what the old state is?

I get frustrated when I have to restart my computer (it sometimes crashes on a re-dock... yay!) and lose the state I've been working on.  You can restart the containers and it will resume if you do it right so they end up with the right IP addresses...  I also had a couple of named (using $OPENSHIFT_CLUSTER_ID) clusters for different things I was fiddling with, and when I flip between tasks, my computer doesn't really have the grunt to run too many clusters.

> At least, maybe "resume" should warn you if the cluster currently has a build that doesn't match the currently-checked-out branch?

Maybe... but I just name the clusters by branch.

> > The new 'ps' command lists the status of all docker containers in the
> > cluster.

> what's the use case for that?

To more easily see the parts of your current cluster if you have multiple named ones.

> > 'move-image' loads an image from the host docker into all node docker
> > daemons.

> Nitpick, but shouldn't that be "copy-image"?

It should.  Changed.

danwinship commented:
> can you split the fixes of existing code/comments into a separate commit?

(I'd still like this)

> It's not the speed, it's the state that I care about.

ah, makes sense

> > Nitpick, but shouldn't that be "copy-image"?
> 
> It should. Changed.

Still says "move-image" twice in the commit message

Otherwise LGTM

knobunc commented:
@danwinship changed as requested.

danwinship commented on hack/dind-cluster.sh:
```diff
@@ -133,20 +133,9 @@ function stop() {
-  fi
+  for cid in $( ${DOCKER_CMD} ps -qa --filter "name=${MASTER_NAME}|${NODE_PREFIX}" ); do
+    ${DOCKER_CMD} rm -f "${cid}" > /dev/null
+  done
```
this should be in the cleanup commit shouldn't it?

danwinship commented on hack/dind-cluster.sh:
```diff
@@ -127,12 +126,84 @@ cluster's rc file to configure the bash environment:
+  local node_name=$3
+
+  echo "Removing node '${node_name}' from dind cluster '${cluster_id}'"
+  sudo echo -n
```
huh? (well, ok, so it's to ensure that we ask for a password sooner rather than later, but why?)

knobunc replied:
Because the state can get into an interesting and hard-to-recover from place if you stop the docker processes and then abort the cleanup that requires sudo.

danwinship commented on hack/dind-cluster.sh:
```diff
@@ -142,6 +213,34 @@ function stop() {
+  # Make a temp file with a descriptor attached
+  tmpfile=$(mktemp /tmp/docker-image.XXXXXX)
+  exec 3>"$tmpfile"
+  rm "$tmpfile"
```
the temporary fd thing seems gratuitous... especially since you're not actually managing to avoid having a tmp file, you're just having it around for less time.

you could use ${config_root} as your temp dir if you don't like /tmp.

knobunc replied:
The fd thing is not to avoid the temp file, and it's not for security (hopefully the mktemp covers that) it's just to make sure we don't leave an image laying around if we fail later.

knobunc commented:
[test]

knobunc commented:
[test] Docker flake: "Error response from daemon: Untar error on re-exec cmd: fork/exec /proc/self/exe: cannot allocate memory"

knobunc commented:
[merge]

knobunc commented:
[merge]

Last failed with a github pull problem (https://ci.openshift.redhat.com/jenkins/job/merge_pull_request_origin/910/):

```
+ wget https://raw.githubusercontent.com/openshift/origin-ci-tool/master/oct/ansible/oct/callback_plugins/default_with_output_lists.py
--2017-06-06 19:17:42--  https://raw.githubusercontent.com/openshift/origin-ci-tool/master/oct/ansible/oct/callback_plugins/default_with_output_lists.py
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 151.101.32.133
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|151.101.32.133|:443... connected.
HTTP request sent, awaiting response... 500 Internal Server Error
2017-06-06 19:17:56 ERROR 500: Internal Server Error.

++ export status=FAILURE
++ status=FAILURE
+ set +o xtrace
########## FINISHED STAGE: FAILURE: INSTALL ANSIBLE PLUGINS [00h 00m 21s] ##########
```
